packages/starknet-checked-felt/src/lib.rs

```diff
@@ -0,0 +1,221 @@
+// Hex - 800000000000011000000000000000000000000000000000000000000000000
+const FELT_MAX_U256: U256 = U256::from_be_bytes::<32>([
+    8, 0, 0, 0, 0, 0, 0, 17, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+]);
+
+impl TryFrom<&[u8; 32]> for CheckedFelt {
+    type Error = CheckedFeltError;
+
+    fn try_from(value: &[u8; 32]) -> Result<Self, Self::Error> {
+        if U256::from_be_bytes::<32>(*value) > FELT_MAX_U256 {
+            return Err(CheckedFeltError::Overflowing);
+        }
+
+        Ok(CheckedFelt(Felt::from_bytes_be(value)))
+    }
+}
+
+impl TryFrom<&[u8]> for CheckedFelt {
+    type Error = CheckedFeltError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() > 32 {
+            return Err(CheckedFeltError::Overflowing);
+        }
+
+        if U256::from_be_slice(value) > FELT_MAX_U256 {
+            return Err(CheckedFeltError::Overflowing);
+        }
+
+        Ok(CheckedFelt(Felt::from_bytes_be_slice(value)))
+    }
+}
+
+impl TryFrom<U256> for CheckedFelt {
+    type Error = CheckedFeltError;
+
+    fn try_from(value: U256) -> Result<Self, Self::Error> {
+        if value > FELT_MAX_U256 {
+            return Err(CheckedFeltError::Overflowing);
+        }
+
+        Ok(CheckedFelt(Felt::from_bytes_be(&value.to_be_bytes::<32>())))
+    }
+}
+
+impl TryFrom<&str> for CheckedFelt {
+    type Error = CheckedFeltError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        CheckedFelt::from_str(value)
+    }
+}
+
+/// Creates a `CheckedFelt` from a hex string value
+impl FromStr for CheckedFelt {
+    type Err = CheckedFeltError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let felt_hex_str = s.trim_start_matches("0x");
+        let felt_hex_bytes = hex::decode(felt_hex_str)?;
+
+        if U256::from_be_slice(felt_hex_bytes.as_slice())
+            > U256::from_be_bytes(Felt::MAX.to_bytes_be())
+        {
+            return Err(CheckedFeltError::Overflowing);
+        }
+
+        Ok(CheckedFelt(Felt::from_bytes_be_slice(
+            felt_hex_bytes.as_slice(),
+        )))
+    }
+}
+
+impl From<CheckedFelt> for Felt {
+    fn from(value: CheckedFelt) -> Self {
+        value.0
+    }
+}
+
+impl From<Felt> for CheckedFelt {
+    fn from(value: Felt) -> Self {
+        CheckedFelt(value)
+    }
+}
+
+impl std::fmt::Display for CheckedFelt {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Deref for CheckedFelt {
+    type Target = Felt;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::fmt::LowerHex for CheckedFelt {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let val = self.0;
+
+        std::fmt::LowerHex::fmt(&val, f)
+    }
+}
+
+impl AsRef<Felt> for CheckedFelt {
+    fn as_ref(&self) -> &Felt {
+        self.0.as_ref()
+    }
+}
```

Comment:
shouldn't be necessary when Deref is defined

Reply:
I removed it initially, but a starknet rust lib function requires `impl AsRef<Felt>` and uses `.as_ref()` on a felt.
I would prefer to keep this impl, because it would make the ampd code a bit cleaner.
I'm bringing it back, but if you'd still prefer that it's removed, we can do that in the ampd PR.